### ISSUE TITLE
"TypeVar only appears once" error isn't always reported.

### ISSUE THREAD

**Issue by kyle-query** [microsoft/pyright#8625]
**Describe the bug**

Pyright correctly reports when an unbound type variable is used in a return type in many cases. But there appear to be some cases where this error is not reported when it should be.

**Code or Screenshots**

<img width="865" alt="Screenshot 2024-07-31 at 4 22 38 PM" src="https://github.com/user-attachments/assets/290a4fb5-9f01-4527-ba97-9fa43a26c1f6">

<br/><p>The return type for `check4` should result in the same error. The definition for `AnyRecordSpec`, from another file, is given here

<img width="624" alt="Screenshot 2024-07-31 at 4 23 22 PM" src="https://github.com/user-attachments/assets/3f2ae648-d6a8-41d2-a57d-0e9768467dbd">

<br/><p>If more context is needed to reproduce this I can probably grant temporary access to the private repo, or attempt to pull this apart into a standalone reproducible example.

I note that one possibly related issue is that `AnyRecordSpec` being a type alias requires manually putting quotes around types that result in a recursive reference back to `AnyRecordSpec`. It's not possible to avoid these using `from __future__ import annotations` nor explicitly marking it as a type alias with `AnyRecordSpec: TypeAlias = ...` probably due to runtime evaluation rules. I think pyright is working correctly there, but at runtime the type alias `AnyRecord = Mapping[str, AnyRecordSpec[....]` reports that `AnyRecord` doesn't take type arguments (presumably because it doesn't decode the quoted type arguments).

**VS Code extension or command-line**

Using both pylance v2024.7.1 and pyright CLI v1.1.373 & v1.1.374.

**Comment by erictraut** [comment]
This check is not exhaustive, and it intentionally exempts the following cases because they can lead to false positives.
* The type variable has value constraints
* The type variable has an explicit default value (PEP 696)
* The type variable is a ParamSpec
* The type variable is used in a specialization for a generic alias, which may use the type variable multiple times in its "expansion"

If these exemptions apply in your case, then pyright is working as intended, and I don't consider it a bug.

If you think that none of these exemptions apply in your case, please provide a minimal, self-contained code sample that demonstrates the issue.

**Comment by erictraut** [comment]
A type variable with an upper bound doesn't affect this check.

The logic doesn't attempt to determine whether a type alias uses the type variable in multiple places. That's quite difficult and costly to do because it requires recursive expansion of the full type alias (since a type alias can defined in terms of other type aliases, etc.). Instead, the check exempts a type variable from this check if it's used in a type argument for a generic type alias. It sounds like that explains what you're seeing.

This check is not meant to be an exhaustive check. It's simply meant to catch common cases of TypeVar misuse that I've seen trip up developers. As with many checks, it errs on the side of potential false negatives to avoid any false positives.

**Comment by kyle-query** [comment]
Ok, I verified your conclusion by changing the definition of `AnyRecordSpec` to `class AnyRecordSpec(Generic[I, O, U, Pb]): ...` and observed that pyright now reports that "TypeVar `I` only appears once" when checking the code `def check5() -> AnyRecordSpec[I, Any, Any, Any]: ...`. Undoing this so `AnyRecordSpec` is the original type alias definition results in that error not being reported.

I assumed this error was simply another way to say "TypeVar `I` is not bound" because it's not introduced into the scope by the function's arguments and it's not introduced by an enclosing class (`check5` has no arguments and isn't a method of a class). I'm not sure I understand why the definition of the return type is needed at all to determine that condition though, so I think you're talking about checking for a different error.

Is it possible to detect that a TypeVar isn't bound, like in this case, when it's used on the return type of a function or method but hasn't been brought into scope by the function parameters or an enclosing generic class declaration?

**Comment by erictraut** [comment]
I'm not sure what you mean. When you say "not bound", do you mean that it goes unsolved for a specific set of arguments that are passed to the call? Or that it could potentially remain unsolved for some theoretical set of arguments?

It would help if you could provide a concrete example where you think pyright should report an error but doesn't currently.

**Comment by kyle-query** [comment]
Sure, below is an example of the scenario I'm talking about. 

```python

from typing import TypeVar, Callable

I = TypeVar("I")

class X0(Generic[I]): ...
X1 = X0[I]

def check5() -> X0[I]: ... # I is only used once
def check6() -> X1[I]: ...

check7: Callable[[], I]    # I has no meaning in this context
```

In typing this out I'm realizing I probably made some unfounded assumptions about Python's type system. I would expect the error reported at `check7` to also be reported at `check5` and `check6`.

**Comment by erictraut** [comment]
Yes, the use of `I` in `check7` is illegal according to the typing spec. A conformant type checker _must_ emit an error in this case.

The use of `I` in `check5` and `check6` is legal, but it's a bad idea and probably indicative of a misunderstanding of how type variables work.  Type checkers are under no obligation to flag these as errors, and mypy doesn't include any such checks. One could make the argument that this check falls more in the domain of linters rather than type checkers, but it's difficult for a typical linter to detect this. I added the `reportInvalidTypeVarUse` warning partly because I field a lot of questions about type variables, and I wanted to help guide developers in the right direction in cases where there was little or no risk of introducing false positives.

As an example of where `check6` would be legitimate and useful, imagine that `X1` is defined as `type X1[T] = Callable[[T], T]`.
